// ngx-logger/esm2015/lib/custom-logger.service.js
import { Inject, Injectable } from '@angular/core';
import { NGXLogger } from './logger.service';
import { TOKEN_LOGGER_MAPPER_SERVICE } from './mapper/imapper.service';
import { TOKEN_LOGGER_METADATA_SERVICE } from './metadata/imetadata.service';
import { TOKEN_LOGGER_RULES_SERVICE } from './rules/irules.service';
import { TOKEN_LOGGER_SERVER_SERVICE } from './server/iserver.service';
import { TOKEN_LOGGER_WRITER_SERVICE } from './writer/iwriter.service';
import { TOKEN_LOGGER_CONFIG_ENGINE_FACTORY } from './config/iconfig-engine-factory';
import * as i0 from "@angular/core";
// Keeping this to avoid any breaking change for now, this class should be removed later
/**
 * CustomNGXLoggerService is designed to allow users to get a new instance of a logger
 * @deprecated The logger is now fully customisable so this class is now useless
 */
export class CustomNGXLoggerService {
    constructor(configEngineFactory, metadataService, ruleService, mapperService, writerService, serverService) {
        this.configEngineFactory = configEngineFactory;
        this.metadataService = metadataService;
        this.ruleService = ruleService;
        this.mapperService = mapperService;
        this.writerService = writerService;
        this.serverService = serverService;
    }
    /**
     * Create an instance of a logger
     * @deprecated The logger is now fully customisable so this function is now useless, if you want a specific instance of NGXLogger, either use the new keyword or Angular dependency injection
     * @param config
     * @param serverService
     * @param logMonitor
     * @param mapperService
     * @returns
     */
    create(config, serverService, logMonitor, mapperService) {
        const logger = new NGXLogger(config, this.configEngineFactory, this.metadataService, this.ruleService, mapperService || this.mapperService, this.writerService, serverService || this.serverService);
        if (logMonitor) {
            logger.registerMonitor(logMonitor);
        }
        return logger;
    }
}
/** @nocollapse */ CustomNGXLoggerService.ɵfac = function CustomNGXLoggerService_Factory(t) { return new (t || CustomNGXLoggerService)(i0.ɵɵinject(TOKEN_LOGGER_CONFIG_ENGINE_FACTORY), i0.ɵɵinject(TOKEN_LOGGER_METADATA_SERVICE), i0.ɵɵinject(TOKEN_LOGGER_RULES_SERVICE), i0.ɵɵinject(TOKEN_LOGGER_MAPPER_SERVICE), i0.ɵɵinject(TOKEN_LOGGER_WRITER_SERVICE), i0.ɵɵinject(TOKEN_LOGGER_SERVER_SERVICE)); };
/** @nocollapse */ CustomNGXLoggerService.ɵprov = i0.ɵɵdefineInjectable({ token: CustomNGXLoggerService, factory: CustomNGXLoggerService.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(CustomNGXLoggerService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [TOKEN_LOGGER_CONFIG_ENGINE_FACTORY]
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [TOKEN_LOGGER_METADATA_SERVICE]
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [TOKEN_LOGGER_RULES_SERVICE]
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [TOKEN_LOGGER_MAPPER_SERVICE]
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [TOKEN_LOGGER_WRITER_SERVICE]
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [TOKEN_LOGGER_SERVER_SERVICE]
            }] }]; }, null); })();
