// ngx-logger/esm2020/testing/lib/config-engine.mock.mjs
import { Injectable } from '@angular/core';
import { NgxLoggerLevel } from 'ngx-logger';
import * as i0 from "@angular/core";
export class NGXLoggerConfigEngineMock {
    get level() {
        return NgxLoggerLevel.ERROR;
    }
    get serverLogLevel() {
        return NgxLoggerLevel.OFF;
    }
    updateConfig(config) {
    }
    partialUpdateConfig(partialConfig) {
    }
    getConfig() {
        return { level: NgxLoggerLevel.ERROR };
    }
}
/** @nocollapse */ /** @nocollapse */ NGXLoggerConfigEngineMock.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: NGXLoggerConfigEngineMock, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
/** @nocollapse */ /** @nocollapse */ NGXLoggerConfigEngineMock.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: NGXLoggerConfigEngineMock });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: NGXLoggerConfigEngineMock, decorators: [{
            type: Injectable
        }] });
