// ngx-logger/esm2020/lib/config/config-engine.mjs
export class NGXLoggerConfigEngine {
    constructor(config) {
        this.config = this._clone(config);
    }
    /** Get a readonly access to the level configured for the NGXLogger */
    get level() {
        return this.config.level;
    }
    /** Get a readonly access to the serverLogLevel configured for the NGXLogger */
    get serverLogLevel() {
        return this.config.serverLogLevel;
    }
    updateConfig(config) {
        this.config = this._clone(config);
    }
    /** Update the config partially
     * This is useful if you want to update only one parameter of the config
     */
    partialUpdateConfig(partialConfig) {
        // avoid any error if the config is incorrect
        if (!partialConfig) {
            return;
        }
        Object.keys(partialConfig).forEach(configParamKey => {
            this.config[configParamKey] = partialConfig[configParamKey];
        });
    }
    getConfig() {
        return this._clone(this.config);
    }
    // TODO: This is a shallow clone, If the config ever becomes hierarchical we must make this a deep clone
    _clone(object) {
        const cloneConfig = { level: null };
        Object.keys(object).forEach((key) => {
            cloneConfig[key] = object[key];
        });
        return cloneConfig;
    }
}
