// ngx-logger/esm2020/lib/metadata/metadata.service.mjs
import { Injectable, Optional } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class NGXLoggerMetadataService {
    constructor(datePipe) {
        this.datePipe = datePipe;
    }
    computeTimestamp(config) {
        const defaultTimestamp = () => new Date().toISOString();
        if (config.timestampFormat) {
            if (!this.datePipe) {
                console.error('NGXLogger : Can\'t use timeStampFormat because DatePipe is not provided. You need to provide DatePipe');
                return defaultTimestamp();
            }
            else {
                return this.datePipe.transform(new Date(), config.timestampFormat);
            }
        }
        return defaultTimestamp();
    }
    getMetadata(level, config, message, additional) {
        const metadata = {
            level: level,
            additional: additional,
        };
        // The user can send a function
        // This is useful in order to compute string concatenation only when the log will actually be written
        if (message && typeof message === 'function') {
            metadata.message = message();
        }
        else {
            metadata.message = message;
        }
        metadata.timestamp = this.computeTimestamp(config);
        return metadata;
    }
}
/** @nocollapse */ /** @nocollapse */ NGXLoggerMetadataService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: NGXLoggerMetadataService, deps: [{ token: i1.DatePipe, optional: true }], target: i0.ɵɵFactoryTarget.Injectable });
/** @nocollapse */ /** @nocollapse */ NGXLoggerMetadataService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: NGXLoggerMetadataService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: NGXLoggerMetadataService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.DatePipe, decorators: [{
                    type: Optional
                }] }]; } });
