// @ngneat/cashew/esm2020/lib/local-storage/local-storage-versions.mjs
import { Injectable } from '@angular/core';
import { storage } from './local-storage';
import { HttpCacheVersions } from '../versions';
import * as i0 from "@angular/core";
const KEY = `@version`;
function createKey(key) {
    return `${KEY}-${key}`;
}
export class LocalStorageVersionsManager extends HttpCacheVersions {
    has(key) {
        return super.has(createKey(key)) || !!storage.getItem(createKey(key));
    }
    get(key) {
        const cacheValue = super.get(createKey(key));
        if (cacheValue) {
            return cacheValue;
        }
        const value = storage.getItem(createKey(key));
        if (value) {
            super.set(createKey(key), value);
        }
        return super.get(createKey(key));
    }
    set(key, version) {
        storage.setItem(createKey(key), version);
        return super.set(createKey(key), version);
    }
    delete(key) {
        super.delete(key);
        storage.clearItem(key);
        return true;
    }
    clear() {
        super.forEach((_, key) => {
            super.delete(key);
            storage.clearItem(key);
        });
    }
}
LocalStorageVersionsManager.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.10", ngImport: i0, type: LocalStorageVersionsManager, deps: null, target: i0.ɵɵFactoryTarget.Injectable });
LocalStorageVersionsManager.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.10", ngImport: i0, type: LocalStorageVersionsManager });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.10", ngImport: i0, type: LocalStorageVersionsManager, decorators: [{
            type: Injectable
        }] });
