// @ngneat/cashew/esm2020/lib/local-storage/local-storage-ttl.mjs
import { Inject, Injectable } from '@angular/core';
import { HTTP_CACHE_CONFIG } from '../cache-config';
import { DefaultTTLManager, TTLManager } from '../ttl-manager';
import { storage } from './local-storage';
import * as i0 from "@angular/core";
const KEY = `@ttl`;
function createKey(key) {
    return `${KEY}-${key}`;
}
export class LocalStorageTTLManager extends TTLManager {
    constructor(config) {
        super();
        this.config = config;
        this.ttl = new DefaultTTLManager(config);
    }
    isValid(key) {
        const valid = this.ttl.isValid(createKey(key));
        if (valid) {
            return true;
        }
        const localStorageTimeStamp = storage.getItem(createKey(key));
        const validInStorage = localStorageTimeStamp > new Date().getTime();
        if (validInStorage) {
            this.ttl.set(createKey(key), localStorageTimeStamp - new Date().getTime());
        }
        return validInStorage;
    }
    set(key, ttl) {
        const resolveTTL = ttl ?? this.config.ttl;
        storage.setItem(createKey(key), new Date().setMilliseconds(resolveTTL));
        this.ttl.set(createKey(key), resolveTTL);
        return this;
    }
    delete(key) {
        this.ttl.delete(createKey(key));
        storage.clearItem(createKey(key));
        return true;
    }
    clear() {
        this.ttl.forEach((_, key) => {
            this.ttl.delete(key);
            storage.clearItem(key);
        });
    }
}
LocalStorageTTLManager.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.10", ngImport: i0, type: LocalStorageTTLManager, deps: [{ token: HTTP_CACHE_CONFIG }], target: i0.ɵɵFactoryTarget.Injectable });
LocalStorageTTLManager.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.10", ngImport: i0, type: LocalStorageTTLManager });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.10", ngImport: i0, type: LocalStorageTTLManager, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [HTTP_CACHE_CONFIG]
                }] }]; } });
