// @ngneat/cashew/esm2020/lib/local-storage/local-storage-cache.mjs
import { HttpResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { HttpCacheStorage } from '../cache-storage';
import { storage } from './local-storage';
import * as i0 from "@angular/core";
const KEY = `@cache`;
function createKey(key) {
    return `${KEY}-${key}`;
}
export class HttpCacheLocalStorage extends HttpCacheStorage {
    has(key) {
        return super.has(createKey(key)) || !!storage.getItem(createKey(key));
    }
    get(key) {
        const cacheValue = super.get(createKey(key));
        if (cacheValue) {
            return cacheValue;
        }
        const value = storage.getItem(createKey(key));
        if (value) {
            super.set(createKey(key), new HttpResponse(value));
        }
        return super.get(createKey(key));
    }
    set(key, response) {
        storage.setItem(createKey(key), response);
        return super.set(createKey(key), response);
    }
    delete(key) {
        storage.clearItem(createKey(key));
        return super.delete(createKey(key));
    }
    clear() {
        super.forEach((value, key) => {
            super.delete(key);
            storage.clearItem(key);
        });
    }
}
HttpCacheLocalStorage.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.10", ngImport: i0, type: HttpCacheLocalStorage, deps: null, target: i0.ɵɵFactoryTarget.Injectable });
HttpCacheLocalStorage.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.10", ngImport: i0, type: HttpCacheLocalStorage });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.10", ngImport: i0, type: HttpCacheLocalStorage, decorators: [{
            type: Injectable
        }] });
