// bn-ng-idle/fesm2015/bn-ng-idle.mjs
import * as i0 from '@angular/core';
import { Injectable, NgModule } from '@angular/core';
import { Subject, merge, fromEvent, timer } from 'rxjs';

class BnNgIdleService {
    constructor() {
        this.expired$ = new Subject();
    }
    startWatching(timeOutSeconds) {
        this.idle$ = merge(fromEvent(document, 'mousemove'), fromEvent(document, 'click'), fromEvent(document, 'mousedown'), fromEvent(document, 'keypress'), fromEvent(document, 'DOMMouseScroll'), fromEvent(document, 'mousewheel'), fromEvent(document, 'touchmove'), fromEvent(document, 'MSPointerMove'), fromEvent(window, 'mousemove'), fromEvent(window, 'resize'));
        this.timeOutMilliSeconds = timeOutSeconds * 1000;
        this.idleSubscription = this.idle$.subscribe((res) => {
            this.resetTimer();
        });
        this.startTimer();
        return this.expired$;
    }
    startTimer() {
        this.timer$ = timer(this.timeOutMilliSeconds, this.timeOutMilliSeconds).subscribe((res) => {
            this.expired$.next(true);
        });
    }
    resetTimer(timeOutSeconds) {
        this.timer$.unsubscribe();
        if (timeOutSeconds) {
            this.timeOutMilliSeconds = timeOutSeconds * 1000;
        }
        this.startTimer();
    }
    stopTimer() {
        this.timer$.unsubscribe();
        this.idleSubscription.unsubscribe();
    }
}
BnNgIdleService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: BnNgIdleService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
BnNgIdleService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: BnNgIdleService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: BnNgIdleService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return []; } });

class BnNgIdleModule {
}
BnNgIdleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: BnNgIdleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
BnNgIdleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: BnNgIdleModule });
BnNgIdleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: BnNgIdleModule, providers: [BnNgIdleService], imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: BnNgIdleModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [],
                    declarations: [],
                    providers: [BnNgIdleService],
                    exports: []
                }]
        }] });

/*
 * Public API Surface of bn-ng-idle
 */

/**
 * Generated bundle index. Do not edit.
 */

export { BnNgIdleModule, BnNgIdleService };
