// @ngneat/cashew/esm2020/lib/ttl-manager.mjs
import { Inject, Injectable } from '@angular/core';
import { HTTP_CACHE_CONFIG } from './cache-config';
import * as i0 from "@angular/core";
export class TTLManager extends Map {
}
export class DefaultTTLManager extends TTLManager {
    constructor(config) {
        super();
        this.config = config;
    }
    isValid(key) {
        return this.get(key) > new Date().getTime();
    }
    set(key, ttl) {
        return super.set(key, new Date().setMilliseconds(ttl || this.config.ttl));
    }
}
DefaultTTLManager.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.10", ngImport: i0, type: DefaultTTLManager, deps: [{ token: HTTP_CACHE_CONFIG }], target: i0.ɵɵFactoryTarget.Injectable });
DefaultTTLManager.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.10", ngImport: i0, type: DefaultTTLManager });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.10", ngImport: i0, type: DefaultTTLManager, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [HTTP_CACHE_CONFIG]
                }] }]; } });
